File: 2021/day02.py
def exec(commands, part2=False):
    state = { 'h': 0, 'd': 0, 'a': 0 }
    run = run2 if part2 else run1

    for command in commands:
        direction, x = command.split(' ')
        run(state, direction, int(x))

    return state['h'] * state['d']

def run1(state, direction, x):
    if direction == 'forward':
        state['h'] += x
    elif direction == 'down':
        state['d'] += x
    else:
        state['d'] -= x

def run2(state, direction, x):
    if direction == 'forward':
        state['h'] += x
        state['d'] += state['a'] * x
    elif direction == 'down':
        state['a'] += x
    else:
        state['a'] -= x

if __name__ == "__main__":
    commands = open("input/day02", "r").read().strip().split('\n')
    print("Part 1:", exec(commands))
    print("Part 2:", exec(commands, True))

File: 2021/day02.ex
defmodule Solution do
  def exec(commands, acc) do
    state = Enum.reduce(commands, acc, &run/2)
    elem(state, 0) * elem(state, 1)
  end

  defp run({"forward", x}, {h, depth}), do: {h + x, depth}
  defp run({"down", x}, {h, depth}), do: {h, depth + x}
  defp run({"up", x}, {h, depth}), do: {h, depth - x}
  defp run({"forward", x}, {h, d, aim}), do: {h + x, d + aim * x, aim}
  defp run({"down", x}, {h, d, aim}), do: {h, d, aim + x}
  defp run({"up", x}, {h, d, aim}), do: {h, d, aim - x}
end

commands =
  "input/day02"
  |> File.read!()
  |> String.trim()
  |> String.split("\n")
  |> Enum.map(fn cmd ->
    [command, steps] = String.split(cmd)
    {command, String.to_integer(steps)}
  end)

IO.inspect(Solution.exec(commands, {0, 0}), label: "Part 1")
IO.inspect(Solution.exec(commands, {0, 0, 0}), label: "Part 2")
